File: suc/swift/051.swift
import SwiftUI

// MARK: - AppTokens (Translation of Kotlin AppTokens)

struct AppTokens {
    struct Colors {
        static let primary = Color(hex: 0xFF00FFFF) // Cyan
        static let secondary = Color(hex: 0xFF00B4FF) // Light Blue
        static let tertiary = Color(hex: 0xFF8A2BE2) // Blue Violet
        static let background = Color(hex: 0xFF0A0A1A) // Dark Navy
        static let surface = Color(hex: 0xFF141436) // Darker Purple-Blue
        static let surfaceVariant = Color(hex: 0xFF1E1E4A) // Even Darker Purple-Blue
        static let outline = Color(hex: 0xFF2E2E5C) // Dark Purple-Blue
        static let success = Color(hex: 0xFF00FFAA) // Greenish Cyan
        static let warning = Color(hex: 0xFFFFCC00) // Yellow
        static let error = Color(hex: 0xFFFF4444) // Red
        static let onPrimary = Color(hex: 0xFF0A0A1A) // Dark Navy (on primary)
        static let onSecondary = Color(hex: 0xFFFFFFFF) // White
        static let onBackground = Color(hex: 0xFFFFFFFF) // White
        static let onSurface = Color(hex: 0xFFFFFFFF) // White
    }

    struct TypographyTokens {
        // Kotlin's sp directly maps to points in SwiftUI for basic scaling
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular) // Kotlin's FontWeight.Normal is .regular in Swift
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        // Kotlin's Dp directly maps to CGFloat for corner radius
        static let smallCornerRadius: CGFloat = 8
        static let mediumCornerRadius: CGFloat = 14
        static let largeCornerRadius: CGFloat = 22
    }

    struct Spacing {
        // Kotlin's Dp directly maps to CGFloat for spacing
        static let sm: CGFloat = 6
        static let md: CGFloat = 10
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 36
    }

    struct ShadowSpec {
        let elevation: CGFloat // Corresponds to Compose's Dp
        let radius: CGFloat    // Corresponds to Compose's Dp for blur radius
        let dy: CGFloat        // Corresponds to Compose's Dp for y-offset
        let opacity: Double    // Corresponds to Compose's Float
    }

    struct ElevationMapping {
        // Shadow parameters are directly mapped
        static let level1 = ShadowSpec(elevation: 2, radius: 4, dy: 2, opacity: 0.1)
        static let level2 = ShadowSpec(elevation: 6, radius: 8, dy: 4, opacity: 0.15)
        static let level3 = ShadowSpec(elevation: 10, radius: 12, dy: 6, opacity: 0.18)
    }
}

// MARK: - Color Extension for Hex Initialization

extension Color {
    /// Initializes a Color from a 32-bit hexadecimal integer (e.g., 0xFFRRGGBB).
    init(hex: UInt) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: 1.0
        )
    }
}

// MARK: - RootScreen (Translation of Kotlin RootScreen Composable)

struct RootScreen: View {
    let stations = ["A", "B", "C", "D", "E", "F"]
    // Standard Material Design BottomAppBar height
    let bottomBarHeight: CGFloat = 56 
    // Standard Material Design FloatingActionButton size
    let fabSize: CGFloat = 56

    var body: some View {
        // ZStack is used to layer the main content, Floating Action Button, and Bottom App Bar,
        // similar to how Scaffold manages these elements in Compose.
        ZStack {
            // Main content area, equivalent to Scaffold's content block.
            // It's a VStack to arrange elements vertically.
            VStack(spacing: AppTokens.Spacing.lg) {
                Text("Metro Planner")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.primary)
                    .frame(maxWidth: .infinity, alignment: .leading) // Equivalent to Modifier.fillMaxWidth() and Alignment.Start

                // Canvas for drawing the metro line and stations.
                // The drawing logic is directly translated from Kotlin's Canvas.
                Canvas { context, size in
                    var path = Path()
                    let step = size.width / CGFloat(stations.count - 1)

                    for (i, _) in stations.enumerated() {
                        let x = CGFloat(i) * step
                        let y = i % 2 == 0 ? size.height / 4 : size.height * 3 / 4
                        if i == 0 {
                            path.move(to: CGPoint(x: x, y: y))
                        } else {
                            path.addLine(to: CGPoint(x: x, y: y))
                        }
                    }

                    // Draw the line with a horizontal gradient brush and stroke width.
                    context.stroke(
                        path,
                        with: .linearGradient(
                            Gradient(colors: [AppTokens.Colors.secondary, AppTokens.Colors.primary]),
                            startPoint: .zero,
                            endPoint: CGPoint(x: size.width, y: 0)
                        ),
                        lineWidth: 6
                    )

                    // Draw the circles for stations.
                    for (i, _) in stations.enumerated() {
                        let x = CGFloat(i) * step
                        let y = i % 2 == 0 ? size.height / 4 : size.height * 3 / 4
                        let center = CGPoint(x: x, y: y)

                        // Outer circle (radius 16f in Kotlin means diameter 32)
                        context.fill(
                            Path(ellipseIn: CGRect(x: center.x - 16, y: center.y - 16, width: 32, height: 32)),
                            with: .color(AppTokens.Colors.primary)
                        )
                        // Inner circle (radius 6f in Kotlin means diameter 12)
                        context.fill(
                            Path(ellipseIn: CGRect(x: center.x - 6, y: center.y - 6, width: 12, height: 12)),
                            with: .color(AppTokens.Colors.background)
                        )
                    }
                }
                .padding(AppTokens.Spacing.md) // Inner padding for the drawing content
                .frame(height: 260) // Fixed height for the canvas
                .background(AppTokens.Colors.surfaceVariant) // Background color for the canvas area
                .cornerRadius(AppTokens.Shapes.largeCornerRadius) // Rounded corners for the canvas background
                .padding(.horizontal, AppTokens.Spacing.lg) // Outer horizontal padding for the entire canvas block


                // Card for "Next Departures" section.
                // Implemented as a VStack with background, corner radius, and shadow.
                VStack(alignment: .leading, spacing: AppTokens.Spacing.sm) {
                    Text("Next Departures")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(AppTokens.Colors.secondary)

                    ForEach(["08:45 • Line 1", "09:10 • Line 2", "09:30 • Line 3"], id: \.self) { departure in
                        HStack { // Equivalent to Row in Compose
                            Text(departure)
                                .font(AppTokens.TypographyTokens.body)
                                .foregroundColor(AppTokens.Colors.onSurface)
                            Spacer() // Equivalent to Modifier.weight(1f) or Arrangement.SpaceBetween
                            Text("On Time")
                                .font(AppTokens.TypographyTokens.label)
                                .foregroundColor(AppTokens.Colors.success)
                        }
                    }
                }
                .padding(AppTokens.Spacing.lg) // Inner padding for the card content
                .background(AppTokens.Colors.surface) // Card background color
                .cornerRadius(AppTokens.Shapes.largeCornerRadius) // Rounded corners for the card
                // Shadow applied based on AppTokens.ElevationMapping.level2
                .shadow(color: AppTokens.Colors.outline.opacity(AppTokens.ElevationMapping.level2.opacity),
                        radius: AppTokens.ElevationMapping.level2.radius,
                        x: 0,
                        y: AppTokens.ElevationMapping.level2.dy)
                .padding(.horizontal, AppTokens.Spacing.lg) // Outer horizontal padding for the entire card block

                // Spacer to push content up, equivalent to Spacer(modifier = Modifier.height(AppTokens.Spacing.md))
                Spacer(minLength: AppTokens.Spacing.md)
            }
            .padding(.top, AppTokens.Spacing.lg) // Top padding for the entire content column
            // Add bottom padding to prevent content from being obscured by the bottom bar.
            // In Compose Scaffold, content padding accounts for bottom bar.
            .padding(.bottom, bottomBarHeight) 
            .frame(maxWidth: .infinity, maxHeight: .infinity) // Equivalent to Modifier.fillMaxSize()
            // Apply vertical gradient background to the main content area.
            .background(
                LinearGradient(
                    gradient: Gradient(colors: [AppTokens.Colors.background, AppTokens.Colors.surfaceVariant, AppTokens.Colors.background]),
                    startPoint: .top,
                    endPoint: .bottom
                )
            )
            // Make the content ignore the safe area to fill the entire screen, including under the status bar.
            .ignoresSafeArea(.all, edges: .all)

            // Floating Action Button (FAB)
            // Positioned using a VStack and HStack with Spacers to push it to the bottom-trailing corner.
            VStack {
                Spacer() // Pushes the FAB to the bottom
                HStack {
                    Spacer() // Pushes the FAB to the trailing edge
                    Button(action: {}) { // Button action is empty as in Kotlin
                        Text("+")
                            .font(AppTokens.TypographyTokens.title)
                            .foregroundColor(AppTokens.Colors.onPrimary)
                            .frame(width: fabSize, height: fabSize) // Fixed size for FAB
                            .background(AppTokens.Colors.primary) // FAB background color
                            .clipShape(Circle()) // Circular shape for FAB
                    }
                    .padding(.trailing, AppTokens.Spacing.lg) // Right padding from screen edge
                    // Bottom padding: FAB is AppTokens.Spacing.lg above the bottom bar.
                    .padding(.bottom, bottomBarHeight + AppTokens.Spacing.lg) 
                }
            }
            .ignoresSafeArea(.all, edges: .all) // FAB also ignores safe area for precise positioning

            // Bottom App Bar
            // Implemented as a VStack containing an HStack, pushed to the bottom of the ZStack.
            VStack(spacing: 0) {
                Spacer() // Pushes the bottom bar to the very bottom
                HStack(spacing: 0) { // Use spacing: 0 and then padding for internal spacing
                    Text("Routes")
                        .font(AppTokens.TypographyTokens.body)
                        .foregroundColor(AppTokens.Colors.secondary) // Active item color
                    Spacer()
                    Text("Planner")
                        .font(AppTokens.TypographyTokens.body)
                        .foregroundColor(AppTokens.Colors.onSurface)
                    Spacer()
                    Text("Tickets")
                        .font(AppTokens.TypographyTokens.body)
                        .foregroundColor(AppTokens.Colors.onSurface)
                    Spacer()
                    Text("Profile")
                        .font(AppTokens.TypographyTokens.body)
                        .foregroundColor(AppTokens.Colors.onSurface)
                }
                .padding(.horizontal, AppTokens.Spacing.lg) // Horizontal padding for the text items
                .frame(maxWidth: .infinity) // Equivalent to Modifier.fillMaxWidth()
                .frame(height: bottomBarHeight) // Fixed height for the bottom bar
                .background(AppTokens.Colors.surface) // Bottom bar background color
            }
            // Make the bottom bar ignore the safe area at the bottom to extend to the very edge.
            .ignoresSafeArea(.all, edges: .bottom)
        }
        .background(AppTokens.Colors.background) // Overall background color if ZStack content doesn't fill
        .statusBarHidden(true) // Hide the status bar for a full-screen experience
    }
}

// MARK: - App Entry Point

// The @main attribute marks MetroPlannerApp as the entry point of the SwiftUI application.
@main
struct MetroPlannerApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}

File: suc/kotlin/051.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.BottomAppBar
import androidx.compose.material3.Card
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import kotlin.random.Random

private const val NAME = "068*MetroPlanner*en"
private const val UI_TYPE = "Travel"
private const val STYLE_THEME = "Dark Neon"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF00FFFF)
        val secondary = Color(0xFF00B4FF)
        val tertiary = Color(0xFF8A2BE2)
        val background = Color(0xFF0A0A1A)
        val surface = Color(0xFF141436)
        val surfaceVariant = Color(0xFF1E1E4A)
        val outline = Color(0xFF2E2E5C)
        val success = Color(0xFF00FFAA)
        val warning = Color(0xFFFFCC00)
        val error = Color(0xFFFF4444)
        val onPrimary = Color(0xFF0A0A1A)
        val onSecondary = Color(0xFFFFFFFF)
        val onBackground = Color(0xFFFFFFFF)
        val onSurface = Color(0xFFFFFFFF)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(8.dp)
        val medium = RoundedCornerShape(14.dp)
        val large = RoundedCornerShape(22.dp)
    }
    object Spacing {
        val sm = 6.dp
        val md = 10.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 36.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.1f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.15f)
        val level3 = ShadowSpec(10.dp, 12.dp, 6.dp, 0.18f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@Composable
fun RootScreen() {
    val stations = remember {
        listOf("A", "B", "C", "D", "E", "F")
    }
    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        floatingActionButton = {
            FloatingActionButton(
                onClick = {},
                containerColor = AppTokens.Colors.primary,
                contentColor = AppTokens.Colors.onPrimary,
                shape = CircleShape
            ) { Text("+", style = AppTokens.TypographyTokens.title) }
        },
        bottomBar = {
            BottomAppBar(
                containerColor = AppTokens.Colors.surface,
                contentColor = AppTokens.Colors.primary
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = AppTokens.Spacing.lg),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text("Routes", color = AppTokens.Colors.secondary, style = AppTokens.TypographyTokens.body)
                    Text("Planner", color = AppTokens.Colors.onSurface, style = AppTokens.TypographyTokens.body)
                    Text("Tickets", color = AppTokens.Colors.onSurface, style = AppTokens.TypographyTokens.body)
                    Text("Profile", color = AppTokens.Colors.onSurface, style = AppTokens.TypographyTokens.body)
                }
            }
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(
                    Brush.verticalGradient(
                        listOf(
                            AppTokens.Colors.background,
                            AppTokens.Colors.surfaceVariant,
                            AppTokens.Colors.background
                        )
                    )
                )
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.lg)
        ) {
            Text("Metro Planner", style = AppTokens.TypographyTokens.display, color = AppTokens.Colors.primary)
            Canvas(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(260.dp)
                    .background(AppTokens.Colors.surfaceVariant, AppTokens.Shapes.large)
                    .padding(AppTokens.Spacing.md)
            ) {
                val path = Path()
                val step = size.width / (stations.size - 1)
                stations.forEachIndexed { i, _ ->
                    val x = i * step
                    val y = if (i % 2 == 0) size.height / 4 else size.height * 3 / 4
                    if (i == 0) path.moveTo(x, y) else path.lineTo(x, y)
                }
                drawPath(
                    path,
                    Brush.horizontalGradient(
                        listOf(AppTokens.Colors.secondary, AppTokens.Colors.primary)
                    ),
                    style = androidx.compose.ui.graphics.drawscope.Stroke(width = 6f)
                )
                stations.forEachIndexed { i, s ->
                    val x = i * step
                    val y = if (i % 2 == 0) size.height / 4 else size.height * 3 / 4
                    drawCircle(color = AppTokens.Colors.primary, radius = 16f, center = Offset(x, y))
                    drawCircle(color = AppTokens.Colors.background, radius = 6f, center = Offset(x, y))
                }
            }
            Card(
                shape = AppTokens.Shapes.large,
                colors = androidx.compose.material3.CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                elevation = androidx.compose.material3.CardDefaults.cardElevation(AppTokens.ElevationMapping.level2.elevation),
                modifier = Modifier.fillMaxWidth()
            ) {
                Column(
                    modifier = Modifier.fillMaxWidth().padding(AppTokens.Spacing.lg),
                    verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)
                ) {
                    Text("Next Departures", style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.secondary)
                    listOf("08:45 • Line 1", "09:10 • Line 2", "09:30 • Line 3").forEach {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween
                        ) {
                            Text(it, style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.onSurface)
                            Text("On Time", style = AppTokens.TypographyTokens.label, color = AppTokens.Colors.success)
                        }
                    }
                }
            }
            Spacer(modifier = Modifier.height(AppTokens.Spacing.md))
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFF0A0A1A)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}
